Add support for float8_e4m3fnuz and _e5m2fnuz

galexite opened the pull request:
This PR relates to the feature in [this feature submission](https://docs.google.com/document/d/1pF2T1xz54IPg1jG7FhykbrpbcJZVelQw0v8vBaoLkfs/edit). It has been based on #104242 which adds similar float8 types.

These new types added in this PR are described in the paper at https://arxiv.org/abs/2206.02915. A brief description and comparison of the types with other float8 types can be also found in the [OpenXLA RFC](https://github.com/openxla/stablehlo/blob/main/rfcs/20230321-fp8_fnuz.md).

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @EikanWang @albanD

galexite committed:
Add support for float8_e4m3fnuz and _e5m2fnuz

This PR relates to the feature specification given in
https://docs.google.com/document/d/1pF2T1xz54IPg1jG7FhykbrpbcJZVelQw0v8vBaoLkfs/edit.

These types are described in the paper at
https://arxiv.org/abs/2206.02915. A brief description and comparison of
the types with other float8 types can be also found in the OpenXLA RFC:
https://github.com/openxla/stablehlo/blob/main/rfcs/20230321-fp8_fnuz.md

Additionally, support for `torch.finfo` for all float8 types, including
existing `e5m2` and `e4m3fn` implementations, has been enabled.

galexite committed:
Lint

galexite committed:
Enable `copy_` and `compare_eq_ne_cuda` for new types

galexite committed:
Add support to `fill_cuda`

galexite commented:
@albanD @seemethere – would it be possible to get an exception for the 2000 LOC limit? I'm only 6 lines over and quite a lot of it is just registering the types.

galexite commented:
I've removed my changes to `finfo` to fit inside the LOC limit – I'll commit them as a separate PR.

vkuzo commented:
Thanks for working on this! At a high level this looks good, and we will need to do a more detailed review.  Given that the branch cut for v2.1 is only a few days away and the high LOC of this change and our experience with landing previous "new dtype" PRs, I would expect that Pytorch v2.2 would be a reasonable target for eventually getting this in.

vkuzo commented on c10/util/Float8_e4m3fnuz.h:
```diff
@@ -0,0 +1,254 @@
+ *
+ * @note The implementation doesn't use any floating-point operations.
+ */
+inline C10_HOST_DEVICE float fp8e4m3fnuz_to_fp32_value(uint8_t input) {
```
this looks reasonable.  Just curious, what made you choose a LUT over bit shifting?

also, do we expect the hardware to support an accelerated version of these?

galexite replied:
There's no particular reason to use a LUT here, I can change to bit shifting if needed!

Yep, Graphcore's C600 hardware has dedicated instructions for which can be used to convert to and from both these FP8 types.

malfet replied:
Did we run any perf benchmarks/binary size consideration for having this lookup table embedded over and over in every op that will need to convert F8E4M3FNUZ to float?

vkuzo commented on test/quantization/core/experimental/test_float8.py:
```diff
@@ -91,7 +128,15 @@ class TestFloat8Dtype(TestCase):
 
-    @parametrize("dtype", [torch.float8_e5m2, torch.float8_e4m3fn])
+    @parametrize(
+        "dtype",
```
nit: maybe something like

```
FLOAT8_DTYPES = [...]

@parametrize("dtype", FLOAT8_DTYPES)
```

to save some repetitive LOC?

vkuzo commented:
The dtype pieces looks reasonable to me!  Wondering if you could share some context on which hardware supports these float8 flavors now, and which hardware is expected to support this in the future?

cc @malfet , would you be up for a more detailed review on the framework pieces figuring out the best way to land this?

galexite commented:
Thanks for the review @vkuzo!

> The dtype pieces looks reasonable to me! Wondering if you could share some context on which hardware supports these float8 flavors now, and which hardware is expected to support this in the future?

Graphcore's [C600 card](https://www.graphcore.ai/products/c600) support these types at the hardware level and these currently available on the market. It has instructions in the [Tile ISA](https://docs.graphcore.ai/projects/isa-mk2-with-fp8/en/latest/_static/TileVertexISA-IPU21-1.3.1.pdf) to perform common operations directly on FP8 data, as well as convert between types.

I'll make the change to take those list of dtypes for the test parameters out in to a constant.

galexite committed:
Extract `FLOAT8_DTYPES` as a constant

galexite commented:
Yeah, it looks like it has gone over. If that's not acceptable, I'm happy to remove it and submit separately.

galexite commented:
@albanD, @seemethere: would it be possible to make an exception for this diff? I'm 42 lines over – I felt leaving the `torch.finfo` support in makes sense as the information required is already in the diff.

seemethere reviewed (approved):
LGTM

galexite committed:
Merge remote-tracking branch 'upstream/main' into georgew/upstream_fp8

galexite commented:
@malfet, may I trouble you for a review please? Thanks

malfet reviewed (changes requested):
This PR feels very much like a copy-n-paste of https://github.com/pytorch/pytorch/pull/104242 that adds 2 more very similar data types.
Which looks fine on the surface, but IMO duplicates the code for F8 data types and leaves lots of questions unanswered, for example how how dtype conversion should work between two f8_e5m2 flavors.

Perhaps more importantly, what is the goal of this PR?  https://github.com/pytorch/pytorch/pull/104242 was a stepping stone for enabling F8 support in ARM cpus and NVIDIA H100 GPUs, but it looks like NZ flavors are only available in GraphCore AI accelerators. If this is the case, then perhaps it would be sufficient to just use `torch.bits8` dtype or indeed add a new enum, but skip dispatch implementation for the basic OPs?

Also, would it be possible to share some of the implementation between say F8_E5M2 and F8_E5M2NZ?

galexite committed:
Merge remote-tracking branch 'upstream/main' into georgew/upstream_fp8

galexite commented on aten/src/ATen/AccumulateType.h:
```diff
@@ -106,7 +108,9 @@ CUDA_ACC_TYPE(half, float);
 CUDA_ACC_TYPE(BFloat16, float);
 CUDA_ACC_TYPE(Half, float);
 CUDA_ACC_TYPE(Float8_e5m2, float);
+CUDA_ACC_TYPE(Float8_e5m2fnuz, float);
```
Both of these were a bad merge on my part – they are not necessary. I've removed them.

galexite committed:
Remove accumulate type on MPS/CUDA for fnuz types

galexite committed:
Merge remote-tracking branch 'upstream/main' into georgew/upstream_fp8

galexite commented:
Thanks for the review @malfet!

> This PR feels very much like a copy-n-paste of #104242 that adds 2 more very similar data types. Which looks fine on the surface, but IMO duplicates the code for F8 data types

Which code in particular? The operator definitions defined in `Float8_e{5m2,4m3}fnuz-inl.h` especially are quite large in LOC and could easily be covered by macros shared between all the float8 types, but they were already duplicated between `Float8_e5m2-inl.h` and `Float8_e4m3fn-inl.h`, so I chose to do the same.

> and leaves lots of questions unanswered, for example how how dtype conversion should work between two f8_e5m2 flavors.

I'm not sure how common it would be to convert between these dtypes, especially if hardware only supports one or the other – I'm happy to provide casts however.

> Perhaps more importantly, what is the goal of this PR? #104242 was a stepping stone for enabling F8 support in ARM cpus and NVIDIA H100 GPUs, but it looks like NZ flavors are only available in GraphCore AI accelerators. If this is the case, then perhaps it would be sufficient to just use `torch.bits8` dtype or indeed add a new enum, but skip dispatch implementation for the basic OPs?

The goal of this PR is to have native support for fnuz types as supported by hardware inside PyTorch. Ideally, this would involve some support for basic ops on CPU (add, mul, copy_, ne, etc.) in the same manner as they are for e5m2, e4m3.

> but it looks like NZ flavors are only available in GraphCore AI accelerators

Support for these types [have been added to the AMDGPU backend of LLVM](https://github.com/llvm/llvm-project/commit/22f0c7a45149a55e643665c4cb69faa960e3c565) for native hardware support on AMD's `gfx940` architecture.

> Also, would it be possible to share some of the implementation between say F8_E5M2 and F8_E5M2NZ?

I'm not sure what you mean by 'share some of the implementation between' – the C600 does not support e5m2, only e5m2fnuz, likewise for e4m2.

galexite committed:
Merge branch 'main' of github.com:pytorch/pytorch

galexite committed:
Remove operators from -inl.h

galexite committed:
Remove gemm/addmm

galexite committed:
Merge branch 'main' of github.com:pytorch/pytorch into georgew/upstream_fp8

galexite commented:
@malfet: I hope I've responded to your questions well enough! Is there a chance you could re-review?

galexite committed:
Merge branch 'main' of github.com:pytorch/pytorch into georgew/upstream_fp8

galexite committed:
Remove even more changes

galexite committed:
Merge branch 'main' of github.com:pytorch/pytorch into georgew/upstream_fp8

galexite committed:
Continued clean-up

galexite committed:
Merge branch 'main' of github.com:pytorch/pytorch into georgew/upstream_fp8

galexite committed:
Merge branch 'main' of github.com:pytorch/pytorch into georgew/upstream_fp8

jeffdaily commented:
> @malfet: I hope I've responded to your questions well enough! Is there a chance you could re-review?

@malfet I too request a re-review.  As pointed out in prior comments, these types are necessary for our gfx940.

malfet commented:
@jeffdaily , @galexite this PR fails 16+ tests because it's trying to run something with CUDA using new dtypes, which:
- Renders it unmergeable
- Not something we want to run/test/compile for, don't we?

malfet commented on aten/src/ATen/AccumulateType.cpp:
```diff
@@ -15,7 +15,7 @@ c10::ScalarType toAccumulateType(c10::ScalarType type, c10::DeviceType device) {
       }
 
-    AT_FORALL_SCALAR_TYPES_WITH_COMPLEX_EXCEPT_COMPLEX_HALF(DEFINE_CASE)
+    AT_FORALL_SCALAR_TYPES_WITH_COMPLEX_EXCEPT_COMPLEX_HALF_F8NZ(DEFINE_CASE)
```
Can you please explain why this one is needed?

galexite replied:
I changed the name of the `AT_FORALL_SCALAR_TYPES_WITH_COMPLEX_EXCEPT_COMPLEX_HALF` macro to make it clear it isn't all scalar types. Unfortunately, were I to add F8NZ types to that macro, I'd have to add CUDA_ACC_TYPEs and MPS_ACC_TYPEs for these.

malfet commented on aten/src/ATen/cpu/vec/vec_base.h:
```diff
@@ -73,7 +73,9 @@ struct is_floating_point:
       std::is_same<T, at::BFloat16>::value ||
       std::is_same<T, at::Float8_e5m2>::value ||
-      std::is_same<T, at::Float8_e4m3fn>::value> {
+      std::is_same<T, at::Float8_e5m2fnuz>::value ||
+      std::is_same<T, at::Float8_e4m3fn>::value ||
+      std::is_same<T, at::Float8_e4m3fnuz>::value> {
```
I think it's safe to remove all float8 dtypes from here for now, as there shouln't be any vec256 instantiations (I can do it in a separate PR though)

malfet commented on aten/src/ATen/native/CPUBlas.cpp:
```diff
@@ -581,7 +581,7 @@ void gemm_batched_with_stride(
       scalar_t *c, int64_t ldc, int64_t batch_stride_c);
 
-AT_FORALL_SCALAR_TYPES_WITH_COMPLEX_EXCEPT_COMPLEX_HALF(INSTANTIATE_BATCHED_GEMM)
+AT_FORALL_SCALAR_TYPES_WITH_COMPLEX_EXCEPT_COMPLEX_HALF_F8NZ(INSTANTIATE_BATCHED_GEMM)
```
Again curious, why this change is needed here?

malfet commented on aten/src/ATen/native/cuda/CompareEQKernel.cu:
```diff
@@ -28,12 +28,22 @@ struct CompareEqFunctor{
-    opmath_symmetric_gpu_kernel_with_scalars<scalar_t, bool>(
-        iter, CompareEqFunctor<scalar_t>(op));
-  });
+C10_NOINLINE void compare_eq_ne_kernel(TensorIteratorBase& iter, EqOpType op) {
+  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND8(
+      kComplexHalf,
+      kHalf,
+      kBFloat16,
+      kBool,
+      kFloat8_e4m3fn,
+      kFloat8_e4m3fnuz,
+      kFloat8_e5m2,
+      kFloat8_e5m2fnuz,
+      iter.common_dtype(),
+      "compare_eq_ne_cuda",
+      [&]() {
+        opmath_symmetric_gpu_kernel_with_scalars<scalar_t, bool>(
+            iter, CompareEqFunctor<scalar_t>(op));
+      });
```
If you don't mind, let's leave those until a followup PRs?

malfet commented on aten/src/ATen/native/cuda/Copy.cu:
```diff
@@ -107,8 +107,8 @@ void direct_copy_kernel_cuda(TensorIteratorBase &iter) {
   } else {
-    AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND4(
-        kHalf, kBool, kBFloat16, kComplexHalf,dtype, "copy_", [&] {
+    AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND6(
+        kHalf, kBool, kBFloat16, kComplexHalf, kFloat8_e5m2fnuz, kFloat8_e4m3fnuz, dtype, "copy_", [&] {
```
Same as above, let's leave it until followups

jeffdaily commented:
To fix the CI failures, is it enough to update aten/src/ATen/native/cuda/FillKernel.cu and add the 2 new dispatch types?

galexite committed:
Merge branch 'main' of github.com:pytorch/pytorch into georgew/upstream_fp8

galexite commented:
Sorry @malfet, @jeffdaily, I backed up some incomplete changes to this PR and have been working in my local copy since. I have uploaded the rest which removes all the CUDA kernels and disables the test for CUDA. The tests will still fail, because the `simulateFp8Precision` function does not work for a certain input for e4m3fnuz, and I haven't got round to debugging it yet.

galexite committed:
Remove from is_floating_point in vec_base.h

jeffdaily commented:
@galexite thanks.  CI is now passing.  The only remaining issue is responding to @malfet's question about the LUT possibly bloating the binary size.

jeffdaily commented:
@galexite have you had a chance to respond to @malfet's question?

AnthonyBarbier commented:
> @galexite have you had a chance to respond to @malfet's question?

Sorry, he's away, he'll be back on Monday. Which question are you talking about? (Can't see any unanswered thread?)

jeffdaily commented:
> > @galexite have you had a chance to respond to @malfet's question?
> 
> Sorry, he's away, he'll be back on Monday. Which question are you talking about? (Can't see any unanswered thread?)

There was concern that the use of the lookup table in the header file would bloat binary sizes.  So @galexite said they would move it from the header to the cpp file.

galexite committed:
Move to .cpp file

galexite committed:
Merge branch 'main' of github.com:pytorch/pytorch into georgew/upstream_fp8

galexite committed:
Export these conversion functions

galexite commented:
I moved `c10::detail::fp8{e4m3,e5m2}fnuz_to_fp32_value` to their respective .cpp files, and have exported them using `C10_API`.

galexite committed:
Move C10_API usage to header file

galexite committed:
Merge branch 'main' of github.com:pytorch/pytorch into georgew/upstream_fp8

galexite committed:
Lint fixes

galexite committed:
Add #include <array> to .cpp file

galexite committed:
Revert debugging changes to simulateFp8Precision

galexite committed:
Lint fixes

galexite commented:
The issue with the test is that `simulateFp8Precision` does not properly round values which would be subnormal in the float8 type to the nearest valid subnormal. This is even the case for the existing types, but it appears troublesome values do not get generated in `x` for the test to fail for these existing types. I have added an extra test for subnormals and have been trying to figure out how to simulate them properly.

As an aside, the `assertEqual` should probably have both tolerances set to 0.

galexite committed:
Fix test_float8.py

galexite committed:
Merge branch 'main' of github.com:pytorch/pytorch into georgew/upstream_fp8

galexite committed:
Refactor test_float8.py

galexite committed:
Use signalling NaNs over quiet NaNs

galexite committed:
Improve test_float8.py

The numeric test for round-trip casting of float8 dtypes originally
consisted of generating a 100x100 tensor in the range 0..max.

This change refactors the test, adds further edge cases and fixes
multiple issues with the lower precision simulation which the results of
the cast were checked against.

Set atol=0 and rtol=0 to ensure an exact equality check.

Co-authored-by: Andrew Fitzgibbon <awf@graphcore.ai>

galexite committed:
Merge branch 'main' of github.com:pytorch/pytorch into georgew/upstream_fp8

jeffdaily commented:
@galexite build failures are real.  Missing symbol for both rocm and cuda builds.

`lld: error: undefined symbol: c10::detail::fp8e4m3fnuz_to_fp32_value(unsigned char)`

galexite commented:
> @galexite build failures are real. Missing symbol for both rocm and cuda builds.
> 
> `lld: error: undefined symbol: c10::detail::fp8e4m3fnuz_to_fp32_value(unsigned char)`

Sorry @jeffdaily, I should have posted here. I'm looking in to that today but we had to re-write the tests (and add a load more), which I've pushed up as a separate PR here: https://github.com/pytorch/pytorch/pull/113361. I'll rebase this PR atop that other one when it lands.

galexite committed:
assert(false) on CUDA or ROCm device

galexite commented:
I think I have worked around the linking issue with CUDA and ROCm, there is just an issue with overflows from the e5m2fnuz cast which I need to resolve (those are the test failures just appearing now).

umangyadav commented on c10/util/Float8_e5m2fnuz-inl.h:
```diff
@@ -0,0 +1,97 @@
+    return c10::Float8_e5m2fnuz(0x7B, c10::Float8_e5m2fnuz::from_bits());
+  }
+  static constexpr c10::Float8_e5m2fnuz lowest() {
+    return c10::Float8_e5m2fnuz(0xFB, c10::Float8_e5m2fnuz::from_bits());
```
This violates CPP standard which says `max() == -lowest()` 
`lowest` should return minimum value representable such that there is not `x` which is smaller than lowest ` x< lowest()` 

This should return `0xFF`

umangyadav commented on c10/util/Float8_e5m2fnuz-inl.h:
```diff
@@ -0,0 +1,97 @@
+    return c10::Float8_e5m2fnuz(0x4, c10::Float8_e5m2fnuz::from_bits());
+  }
+  static constexpr c10::Float8_e5m2fnuz max() {
+    return c10::Float8_e5m2fnuz(0x7B, c10::Float8_e5m2fnuz::from_bits());
```
This should be `0x7F` for the FNUZ types. FNUZ types do not have  infinity. 
NaN is represented by negative zero. So `7C`, `7D`, 7E` and `7F` are available in representable range.

galexite committed:
Fix numeric_limits

galexite committed:
Merge branch 'main' of github.com:pytorch/pytorch into georgew/upstream_fp8

galexite commented:
Yeah, there were a couple of issues with the `numeric_limits`, I hope I've fixed all of them.

galexite committed:
Attempt to fix HIP build

galexite committed:
Merge branch 'main' of github.com:pytorch/pytorch into georgew/upstream_fp8

galexite committed:
Add C10_HOST_DEVICE to class members

malfet reviewed (approved):
Thank you very much for a nice refactors, looks good to me.

galexite committed:
__CUDA__ vs __CUDA_ARCH__

galexite committed:
Merge branch 'main' of github.com:pytorch/pytorch into georgew/upstream_fp8

umangyadav commented on c10/util/Float8_e4m3fnuz.h:
```diff
@@ -0,0 +1,163 @@
+   * (i.e. the first value which would overflow in to the sign bit, resulting in
+   * a NaN) in fp8e4m3fnuz range:
+   * 1 0000 000 - fp8e4m3fnuz
+   * 0 10000110 00000000000000000000000 - fp32
```
```suggestion
   * 0 10000111 00000000000000000000000 - fp32
```

umangyadav commented on c10/util/Float8_e4m3fnuz.h:
```diff
@@ -0,0 +1,163 @@
+
+  if (f_bits >= fnuz_max) {
+    // NaN -- sign bit set to 1, rest 0s
+    return 0x80;
```
table here clips more than FNUZ_MAX to FLT_MAX. 
https://onnx.ai/onnx/technical/float8.html#cast 

Is there reason behind using NaNs ?

galexite replied:
The reason is that the existing casting code for `e5m2` and `e4m3fn` types are also implemented without any saturation, so I chose to do the same to match that behaviour.

umangyadav replied:
I fear it would lead to having more NaNs when doing inference.  Ideally there should be a flag for the users to set which behaviour they want

umangyadav commented on c10/util/Float8_e4m3fnuz.h:
```diff
@@ -0,0 +1,163 @@
+    return 0x80;
+  }
+
+  if (f_bits < (UINT32_C(121) << 23)) {
```
where this 121 came from ?

galexite replied:
Fixed and comment added.

alugorey commented on c10/util/Float8_e4m3fnuz.h:
```diff
@@ -0,0 +1,163 @@
+ * representation, to a 32-bit floating-point number in IEEE single-precision
+ * format, in bit representation.
+ *
+ * @note The implementation doesn't use any floating-point operations.
+ */
+#if defined(__CUDA__) || defined(__HIP__)
+C10_HOST_DEVICE C10_API inline float fp8e4m3fnuz_to_fp32_value(uint8_t) {
+#if defined(__CUDA_ARCH__) || defined(__HIP_ARCH__)
+  // TORCH_INTERNAL_ASSERT is not supported on a device
+  assert(false);
+#else
+  TORCH_INTERNAL_ASSERT(false, "e4m3fnuz is not supported by CUDA or HIP");
+#endif
+}
+#else
+C10_API float fp8e4m3fnuz_to_fp32_value(uint8_t input);
+#endif
+
```
This entire section can be simplified to:
```
#if defined(__CUDA_ARCH__) || defined(__HIP__)
C10_HOST_DEVICE C10_API inline float fp8e4m3fnuz_to_fp32_value(uint8_t) {
  CUDA_KERNEL_ASSERT(false);
}
#else
C10_API float fp8e4m3fnuz_to_fp32_value(uint8_t input);
#endif
```
We can't have a naked `assert` in device code so needs to change to `CUDA_KERNEL_ASSERT`. This should get past your ROCM CI failure. change also needs to happen in e5m4fnuz variant

alugorey commented on c10/util/Float8_e4m3fnuz.h:
```diff
@@ -0,0 +1,163 @@
+#include <iosfwd>
+#include <ostream>
+
+#if defined(__CUDA_ARCH__) || defined(__HIP_ARCH__)
+#include <cassert>
+#endif
```
This can be removed. See comment below.

alugorey reviewed (changes requested):
A couple comments to help with your rocm CI failure. Hope this helps! :)

malfet commented on c10/util/Float8_e4m3fnuz.h:
```diff
@@ -0,0 +1,163 @@
+ */
+#if defined(__CUDA__) || defined(__HIP__)
+C10_HOST_DEVICE C10_API inline float fp8e4m3fnuz_to_fp32_value(uint8_t) {
+#if defined(__CUDA_ARCH__) || defined(__HIP_ARCH__)
+  // TORCH_INTERNAL_ASSERT is not supported on a device
+  assert(false);
+#else
+  TORCH_INTERNAL_ASSERT(false, "e4m3fnuz is not supported by CUDA or HIP");
+#endif
```
```suggestion
  CUDA_KERNEL_ASSERT(false && "e4m3fnuz is not supported by CUDA or HIP");
  return -1.0;
#endif
```

malfet commented on c10/util/Float8_e5m2fnuz.h:
```diff
@@ -0,0 +1,162 @@
+#include <iosfwd>
+#include <ostream>
+
+#if defined(__CUDA_ARCH__) || defined(__HIP_ARCH__)
+#include <cassert>
+#endif
+
```
```suggestion
```

malfet commented on c10/util/Float8_e4m3fnuz.h:
```diff
@@ -0,0 +1,163 @@
+#include <iosfwd>
+#include <ostream>
+
+#if defined(__CUDA_ARCH__) || defined(__HIP_ARCH__)
+#include <cassert>
+#endif
+
```
```suggestion
```

malfet commented on c10/util/Float8_e5m2fnuz.h:
```diff
@@ -0,0 +1,162 @@
+ */
+#if defined(__CUDA__) || defined(__HIP__)
+C10_HOST_DEVICE C10_API inline float fp8e5m2fnuz_to_fp32_value(uint8_t) {
+#if defined(__CUDA_ARCH__) || defined(__HIP_ARCH__)
+  // TORCH_INTERNAL_ASSERT is not supported on a device
+  assert(false);
+#else
+  TORCH_INTERNAL_ASSERT(false, "e5m2fnuz is not supported by CUDA or HIP");
+#endif
```
```suggestion
  CUDA_KERNEL_ASSERT(false && "e5m2fnuz is not supported by CUDA or HIP");
  return -1.0
```

galexite committed:
Fixes from comments

galexite committed:
Fix typo with CUDA dtypes in test_float8.py

galexite committed:
Lint

galexite committed:
Do the same to e4m3fnuz

malfet commented:
@pytorchbot merge -i
